### DISCUSSION TITLE
Images without foreground cass in the training

### DISCUSSION THREAD

**Discussion by arthursw** [MIC-DKFZ/nnUNet#1140] [Q&A] [answered]
Is it a problem to have a portion of the training image without any foreground class? 
(patients without lesions in the MS lesion segmentation case)

I think it won't help training the network since the dice score and entropy loss cannot be computed for those cases ; is it correct?
Does that mean we should avoid all cases without any foreground class or could it have any benefit?

**Comment by FabianIsensee** [comment, marked as answer]
Hi, this should not be a problem at all. There might be some stdout warning that you can ignore. I would still include these cases in the training because the model will be more robust if it also sees cases where nothing should be predicted. The dice loss is defined for those cases of course and it will push the network to predict nothing if nothing is present

**Reply by arthursw** [reply]
All right, thank you very much!

**Comment by arthursw** [comment]
Just to make sure I understand correctly:

The dice loss is defined as:
```
        tp, fp, fn, _ = get_tp_fp_fn_tn(x, y, axes, loss_mask, False)

        nominator = 2 * tp + self.smooth
        denominator = 2 * tp + fp + fn + self.smooth

        dc = nominator / (denominator + 1e-8)
```

So for images (or patches) without foreground case, the dice loss will be 0 no matter if there are false positives or not. That is why I said (too quickly) "the dice loss is not defined for those cases" ; I meant it will not be *affected* by those cases.

But looking more closely, I understand that `batch_dice` is set to True (so the dice loss is computed on a batch, not a single patch), meaning if a batch contains a patch with foreground class, and one without, then the dice loss will be affected by the number of false positives. Right?

So if I have too many images without foreground classes, I might end up with batches without foreground class, which would not help the training?

**Reply by arthursw** [reply]
Ok I was wrong, I didn't see the `self.smooth` which prevent the loss from being 0 when there is no foreground class! And `batch_dice` is `False` in my case. Ok, it's all clear now :)
